cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkTransactional.java

```diff
@@ -110,112 +138,75 @@ TransactionInfo getTransactionInfo(String key) {
-   * {@link TransactionFailureException} if it finds one.
-   *
-   * @param runnable The {@link TxRunnable} to be executed inside a transaction
-   * @param asyncCommit if {@code true}, the transaction will be left open and it is expected to be committed
-   *                    by external entity asynchronously after this method returns;
-   *                    if {@code false}, the transaction will be committed when the runnable returned.
-   *
-   */
-  private void executeLong(SparkTxRunnable runnable, boolean asyncCommit) throws TransactionFailureException {
-    TransactionalDatasetContext transactionalDatasetContext = activeDatasetContext.get();
-    if (transactionalDatasetContext != null) {
-      try {
-        if (!transactionalDatasetContext.commitOnJobEnded()) {
-          throw new TransactionFailureException("Nested transaction not supported. Active transaction is "
-                                                  + transactionalDatasetContext.getTransaction());
-        }
-
-        if (transactionalDatasetContext.isJobStarted()) {
-          // Wait for the completion of the active transaction. This is needed because the commit is done
-          // asynchronously through the SparkListener.
-          transactionalDatasetContext.awaitCompletion();
-        } else if (!asyncCommit) {
-          // For implicit transaction (commitOnJobEnd == true), if the job hasn't been started,
-          // we can use that transaction.
-          // If asyncCommit is false, change the transaction to be committed at the end of this method instead
-          // of on job end. This is for the case when there is an opening of implicit transaction (e.g. reduceByKey),
-          // followed by an explicit transaction.
-          transactionalDatasetContext.useSyncCommit();
-          LOG.debug("Converted implicit transaction to explicit transaction: {}",
-                    transactionalDatasetContext.getTransaction().getWritePointer());
+      // If the current transaction is commit on job end, we need some special handling
+      if (currentTransactionType == TransactionType.IMPLICIT_COMMIT_ON_JOB_END) {
+        // If the job already started, we need to wait for the job completion so that the active transaction
+        // gets committed before we start a new one. The waiting will block this thread until the job that
+        // associated with the transaction is completed (asynchronously).
+        if (txDatasetContext.isJobStarted()) {
+          try {
+            txDatasetContext.awaitCompletion();
+            txDatasetContext = null;
+          } catch (InterruptedException e) {
+            // Don't execute the runnable. Reset the interrupt flag and return
+            Thread.currentThread().interrupt();
+            return;
+          }
+        } else if (transactionType != TransactionType.IMPLICIT_COMMIT_ON_JOB_END) {
+          // If the job hasn't been started and the requested type is not commit on job end,
```

Comment:
Something like this:

```scala
sc.fromDataset("xxx")
   .reduceByKey(_ + _)
   .collect()
```

See the javadoc for the `IMPLICIT_COMMIT_ON_JOB_END`